course-definition.yml

```diff
@@ -252,6 +252,10 @@ stages:
                 - `piece length`: number of bytes in each piece
                 - `pieces`: concatenated SHA-1 hashes of each piece
 
+        {{#lang_is_java}}
```

Comment:
@sarp A higher-level problem here is that we haven't enforced this in the starter code. I probably should've thought of this when creating the Go & Rust starters, my bad. It is handled properly for Python, we're using [bytestrings](https://github.com/codecrafters-io/build-your-own-bittorrent/blob/6a7e71b25c75b324609daf5aa8295f2721001748/compiled_starters/python/app/main.py#L9). 

So that would be step 1 - convert the starters to use `byte[]` (or whatever the equivalent is, in the respective language). 

As for this message itself - makes sense to keep it since this is likely the first time a user is going to see non-utf characters. I think we could shorten it a bit, and have a version that applies to non-Java languages too (and refine further for each language if people say it's confusing).

My suggestion would be something like: 

---

**Note**: `.torrent` files contain bytes that aren’t valid UTF-8 characters. You’ll run into problems if you try to read the contents of this file as a `String`. Use `byte[]` instead.

(that's for Java)

---


**Note**: `.torrent` files contain bytes that aren’t valid UTF-8 characters. You might run into problems if you try to read the contents of this file as a string. Depending on the language you're using, you might need to use a byte sequence instead ([bytestrings](https://stackoverflow.com/questions/6224052/what-is-the-difference-between-a-string-and-a-byte-string) in Python, `[]byte` in Go etc.)

(that's for other languages)

Reply:
It's not a problem in Go because string holds arbitrary bytes. It is not required to hold UTF-8 text, or any other predefined format. It's possible to get underlying bytes even if you read it as a string (good explanation here: https://go.dev/blog/strings)

For Rust, I'm seeing complaints about torrent containing non UTF-8 characters. 

Thanks for the message suggestion, will update, way better explanation than mine :)

Reply:
Regarding starter code setting people in the wrong direction:
I think it's fine as is. In stages 1-4, we're only dealing with strings from command line, so it's suitable within that scope. In Stage 5, requirements change, torrent file introduces binary data, so the parser they've created requires refactoring. This happens all the time in real life :) I think this is a more natural transition rather than starting people with a `byte[]` and having them question why ... What do you think?

I also considered adding a binary input to the tester for Stage 1 but I think it would be too early to deal with gnarly details when people aren't that invested in the challenge yet,  so decided not to.

Reply:
> What do you think?

Yep, now that you mention it.... I can see it going both ways - don't think there's a clear answer. Add binary stuff early, people will get put off at early stages, don't add it at all and they might be surprised at stage 5. 

Overall, I think your approach works - we'd basically be choosing to optimize for early stages. If we observe that a lot of people complain at stage 5 (even after adding the instruction change), I suppose we could re-evaluate.

Reply:
sounds good. landed these changes separately in case we had a lengthy discussion here. closing this PR for now
- [warn about random peer id](https://github.com/codecrafters-io/build-your-own-bittorrent/commit/b16cb4a00b95fa362b3bed1466a76b5c10027ce9)
- [for rust and java, add binary content warning in stage 5](https://github.com/codecrafters-io/build-your-own-bittorrent/commit/e013c2ec43c3b6fa4aadf9e6ee1149f4ef5130ad)